src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs

```diff
@@ -168,5 +167,18 @@ private bool CloseHandle(bool abortive, bool canceledOperations)
             }
 #endif
         }
+
+        private new void SetHandle(IntPtr handle)
+        {
+            base.SetHandle(handle);
+
+            if (IsInvalid)
+            {
+                SetHandleAsInvalid();
```

Comment:
This is fine as long as no one ever tries to do:
```C#
SetHandle((IntPtr)(-1));
SetHandle(realHandle);
```
as the SafeHandle will have already been marked as invalid at that point.

I'd suggest adding a `Debug.Assert(!IsClosed)` at the beginning of the method to help flag such cases.